source_ds/mm_main_ds.c

```diff
@@ -19,10 +19,183 @@
  *
  ******************************************************************************/
 
+#include "maxmod.h"
 #include "mm_main.h"
+#include "mp_mas_structs.h"
 #include "mm_main_ds.h"
+#include "mm_comms.h"
+#include "mm_comms7.h"
+#include "mm_mas.h"
+#include "mm_effect.h"
+#include "mp_defs.h"
+#include "devkitpro_defs.h"
+
+#if defined(SYS_GBA)
+#include "mm_main_gba.h"
+#include "mm_mixer_gba.h"
+#elif defined(SYS_NDS)
+#include "mm_main_ds.h"
+#include "mm_mixer_ds.h"
+#include "mm_mixer_super.h"
+#endif
 
 void mmInit7(void);
+void mmInitialize(mm_bool);
+mm_word mmEventForwarder(mm_word, mm_word);
+void StopActiveChannel(mm_word);
+
+#define NUM_CHANNELS 32
+
+#define INITIALIZED_VALUE 42
+#define BASE_VOLUME 0x400
+#define BASE_TEMPO 0x400
+#define BASE_PITCH 0x400
+#define ALL_CHANNELS_UNLOCK 0xFFFF
+
+#define TIMER_0_IRQ (1<<3)
+
+extern mm_byte mm_mixing_mode;
+
+mm_module_channel mm_rds_pchannels[NUM_CHANNELS];
+mm_active_channel mm_rds_achannels[NUM_CHANNELS];
+
+// Returns true if the system is ready for playback
+mm_bool mmIsInitialized(void) {
+    return mmInitialized == INITIALIZED_VALUE;
+}
+
+void mmInitialize(mm_bool do_initialize) {
+    if(do_initialize)
+        mmInitialized = INITIALIZED_VALUE;
+    else
+        mmInitialized = 0;
+}
+
+// Unlock audio channels so they can be used by the sequencer.
+void mmUnlockChannels(mm_word mask) {
+#ifdef SYS_NDS
+    // Can NOT unlock channels in mode b
+	if(mm_mixing_mode == 1)
+	    return;
+#endif
+	mm_ch_mask |= mask;
+}
+
+// Lock audio channels to prevent the sequencer from using them.
+void mmLockChannels(mm_word mask) {
+    mm_ch_mask &= ~mask;
+
+    // It stops them, even if they were already locked
+    for(int i = 0; i < NUM_CHANNELS; i++)
+        if(mask & (1 << i))
+            StopActiveChannel(i);
+}
+
+// Stop active channel
+void StopActiveChannel(mm_word index) {
+    mm_module_channel *channels;
+    mm_word num_channels;
+
+#ifdef SYS_NDS
+    mm_mixer_channel *mix_ch = &mm_mix_channels[index];
+#endif
+#ifdef SYS_GBA
+    mm_mixer_channel *mix_ch = &mm_mixchannels[index];
+#endif
+
+#ifdef SYS_GBA
+    mix_ch->src = -1;
+#endif
+#ifdef SYS_NDS
+    mix_ch->samp_cnt = 0;
+    mix_ch->vol = 0;
+    mix_ch->cvol = 0;
+#endif
+
+	*((mm_word*)(REG_SOUNDXCNT+index)) = 0;
```

Comment:
Awkward formatting; can we have consistent indentation?

Reply:
If you mean the tabs, yeah. Fixed that. The asm files use Tabs, so some can slip by.

Reply:
Thank you!